File: swift-bridge/Sources/AVPlayerBridge/AssetImageGenerator.swift
import AVFoundation
import CoreGraphics
import Foundation

private struct AssetImageGeneratorInfoPayload: Codable {
    let appliesPreferredTrackTransform: Bool
    let maximumSize: SizePayload
    let apertureMode: String?
    let requestedTimeToleranceBefore: TimePayload
    let requestedTimeToleranceAfter: TimePayload
    let dynamicRangePolicy: String?
    let hasCustomVideoCompositor: Bool
    let customVideoCompositorClassName: String?
}

private struct AssetImagePayload: Codable {
    let width: Int
    let height: Int
    let bitsPerComponent: Int
    let bitsPerPixel: Int
    let bytesPerRow: Int
    let alphaInfo: UInt32
    let bitmapInfo: UInt32
    let renderingIntent: UInt32
}

private final class AVPAssetImageBox: NSObject {
    let image: CGImage

    init(image: CGImage) {
        self.image = image
    }
}

private func avpAssetImageGeneratorApertureModeString(
    _ mode: AVAssetImageGenerator.ApertureMode?
) -> String? {
    guard let mode else { return nil }
    switch mode {
    case .cleanAperture:
        return "clean_aperture"
    case .productionAperture:
        return "production_aperture"
    case .encodedPixels:
        return "encoded_pixels"
    default:
        return mode.rawValue
    }
}

private func avpAssetImageGeneratorApertureMode(
    from raw: String
) -> AVAssetImageGenerator.ApertureMode {
    switch raw {
    case "clean_aperture":
        return .cleanAperture
    case "production_aperture":
        return .productionAperture
    case "encoded_pixels":
        return .encodedPixels
    default:
        return AVAssetImageGenerator.ApertureMode(rawValue: raw)
    }
}

@available(macOS 15.0, *)
private func avpAssetImageGeneratorDynamicRangePolicyString(
    _ policy: AVAssetImageGenerator.DynamicRangePolicy?
) -> String? {
    guard let policy else { return nil }
    switch policy {
    case .forceSDR:
        return "force_sdr"
    case .matchSource:
        return "match_source"
    default:
        return policy.rawValue
    }
}

@available(macOS 15.0, *)
private func avpAssetImageGeneratorDynamicRangePolicy(
    from raw: String
) -> AVAssetImageGenerator.DynamicRangePolicy {
    switch raw {
    case "force_sdr":
        return .forceSDR
    case "match_source":
        return .matchSource
    default:
        return AVAssetImageGenerator.DynamicRangePolicy(rawValue: raw)
    }
}

private func makeAssetImageGeneratorInfoPayload(
    _ generator: AVAssetImageGenerator
) -> AssetImageGeneratorInfoPayload {
    AssetImageGeneratorInfoPayload(
        appliesPreferredTrackTransform: generator.appliesPreferredTrackTransform,
        maximumSize: encodeSize(generator.maximumSize),
        apertureMode: avpAssetImageGeneratorApertureModeString(generator.apertureMode),
        requestedTimeToleranceBefore: encodeTime(generator.requestedTimeToleranceBefore),
        requestedTimeToleranceAfter: encodeTime(generator.requestedTimeToleranceAfter),
        dynamicRangePolicy: {
            if #available(macOS 15.0, *) {
                return avpAssetImageGeneratorDynamicRangePolicyString(generator.dynamicRangePolicy)
            }
            return nil
        }(),
        hasCustomVideoCompositor: generator.customVideoCompositor != nil,
        customVideoCompositorClassName: generator.customVideoCompositor.map {
            String(describing: type(of: $0))
        }
    )
}

private func makeAssetImagePayload(_ image: CGImage) -> AssetImagePayload {
    AssetImagePayload(
        width: image.width,
        height: image.height,
        bitsPerComponent: image.bitsPerComponent,
        bitsPerPixel: image.bitsPerPixel,
        bytesPerRow: image.bytesPerRow,
        alphaInfo: UInt32(image.alphaInfo.rawValue),
        bitmapInfo: image.bitmapInfo.rawValue,
        renderingIntent: UInt32(image.renderingIntent.rawValue)
    )
}

@_cdecl("av_asset_image_generator_create")
public func av_asset_image_generator_create(
    _ assetPtr: UnsafeMutableRawPointer
) -> UnsafeMutableRawPointer? {
    let asset = Unmanaged<AVAsset>.fromOpaque(assetPtr).takeUnretainedValue()
    return Unmanaged.passRetained(AVAssetImageGenerator(asset: asset)).toOpaque()
}

@_cdecl("av_asset_image_generator_info_json")
public func av_asset_image_generator_info_json(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    do {
        return ffiString(try avpEncodeJSON(makeAssetImageGeneratorInfoPayload(generator)))
    } catch {
        outErrorMessage?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

@_cdecl("av_asset_image_generator_set_applies_preferred_track_transform")
public func av_asset_image_generator_set_applies_preferred_track_transform(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ applies: Bool
) {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.appliesPreferredTrackTransform = applies
}

@_cdecl("av_asset_image_generator_set_maximum_size")
public func av_asset_image_generator_set_maximum_size(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ width: Double,
    _ height: Double
) {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.maximumSize = CGSize(width: width, height: height)
}

@_cdecl("av_asset_image_generator_set_aperture_mode")
public func av_asset_image_generator_set_aperture_mode(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ modePtr: UnsafePointer<CChar>?
) {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.apertureMode = modePtr.map { avpAssetImageGeneratorApertureMode(from: String(cString: $0)) }
}

@_cdecl("av_asset_image_generator_set_requested_time_tolerance_before")
public func av_asset_image_generator_set_requested_time_tolerance_before(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ value: Int64,
    _ timescale: Int32,
    _ kind: Int32
) {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.requestedTimeToleranceBefore = cmTime(value: value, timescale: timescale, kind: kind)
}

@_cdecl("av_asset_image_generator_set_requested_time_tolerance_after")
public func av_asset_image_generator_set_requested_time_tolerance_after(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ value: Int64,
    _ timescale: Int32,
    _ kind: Int32
) {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.requestedTimeToleranceAfter = cmTime(value: value, timescale: timescale, kind: kind)
}

@_cdecl("av_asset_image_generator_set_dynamic_range_policy")
public func av_asset_image_generator_set_dynamic_range_policy(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ policyPtr: UnsafePointer<CChar>,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    guard #available(macOS 15.0, *) else {
        outErrorMessage?.pointee = ffiString("AVAssetImageGenerator.dynamicRangePolicy requires macOS 15.0")
        return AVP_INVALID_ARGUMENT
    }
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.dynamicRangePolicy = avpAssetImageGeneratorDynamicRangePolicy(from: String(cString: policyPtr))
    return AVP_OK
}

@_cdecl("av_asset_image_generator_copy_image_at_time")
public func av_asset_image_generator_copy_image_at_time(
    _ generatorPtr: UnsafeMutableRawPointer,
    _ value: Int64,
    _ timescale: Int32,
    _ kind: Int32,
    _ outActualValue: UnsafeMutablePointer<Int64>?,
    _ outActualTimescale: UnsafeMutablePointer<Int32>?,
    _ outActualKind: UnsafeMutablePointer<Int32>?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    let requestedTime = cmTime(value: value, timescale: timescale, kind: kind)

    var actualTime = CMTime.invalid
    var image: CGImage?

    guard #available(macOS 13.0, *) else {
        outErrorMessage?.pointee = ffiString("AVAssetImageGenerator.image(at:) requires macOS 13.0")
        return nil
    }

    let status = avpBlockOnAsync(
        work: { try await generator.image(at: requestedTime) },
        onSuccess: { result in
            image = result.image
            actualTime = result.actualTime
        },
        outErrorMessage: outErrorMessage
    )
    guard status == AVP_OK, let image else { return nil }
    writeTime(actualTime, value: outActualValue, timescale: outActualTimescale, kind: outActualKind)
    return Unmanaged.passRetained(AVPAssetImageBox(image: image)).toOpaque()
}

@_cdecl("av_asset_image_generator_cancel_all_image_generation")
public func av_asset_image_generator_cancel_all_image_generation(
    _ generatorPtr: UnsafeMutableRawPointer
) {
    let generator = Unmanaged<AVAssetImageGenerator>.fromOpaque(generatorPtr).takeUnretainedValue()
    generator.cancelAllCGImageGeneration()
}

@_cdecl("av_asset_image_info_json")
public func av_asset_image_info_json(
    _ imagePtr: UnsafeMutableRawPointer,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    let image = Unmanaged<AVPAssetImageBox>.fromOpaque(imagePtr).takeUnretainedValue().image
    do {
        return ffiString(try avpEncodeJSON(makeAssetImagePayload(image)))
    } catch {
        outErrorMessage?.pointee = ffiString(error.localizedDescription)
        return nil
    }
}

private func writeTime(
    _ time: CMTime,
    value: UnsafeMutablePointer<Int64>?,
    timescale: UnsafeMutablePointer<Int32>?,
    kind: UnsafeMutablePointer<Int32>?
) {
    let encoded = encodeTime(time)
    switch encoded.kind {
    case "numeric":
        value?.pointee = encoded.value ?? 0
        timescale?.pointee = encoded.timescale ?? 1
        kind?.pointee = 0
    case "indefinite":
        value?.pointee = 0
        timescale?.pointee = 0
        kind?.pointee = 2
    case "positive_infinity":
        value?.pointee = 0
        timescale?.pointee = 0
        kind?.pointee = 3
    case "negative_infinity":
        value?.pointee = 0
        timescale?.pointee = 0
        kind?.pointee = 4
    default:
        value?.pointee = 0
        timescale?.pointee = 0
        kind?.pointee = 1
    }
}

File: src/asset_image_generator.rs
#![allow(clippy::missing_errors_doc, clippy::must_use_candidate)]

use core::ffi::{c_char, c_void};
use core::ptr;
use std::ffi::CString;

use serde::Deserialize;

use crate::asset::{Asset, Size, UrlAsset};
use crate::error::{from_swift, AVPlayerError};
use crate::ffi;
use crate::time::Time;
use crate::util::parse_json_and_free;

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
struct AssetImageGeneratorInfoPayload {
    applies_preferred_track_transform: bool,
    maximum_size: Size,
    aperture_mode: Option<String>,
    requested_time_tolerance_before: Time,
    requested_time_tolerance_after: Time,
    dynamic_range_policy: Option<String>,
    has_custom_video_compositor: bool,
    custom_video_compositor_class_name: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
struct AssetImagePayload {
    width: usize,
    height: usize,
    bits_per_component: usize,
    bits_per_pixel: usize,
    bytes_per_row: usize,
    alpha_info: u32,
    bitmap_info: u32,
    rendering_intent: u32,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum AssetImageGeneratorApertureMode {
    CleanAperture,
    ProductionAperture,
    EncodedPixels,
    Unknown(String),
}

impl AssetImageGeneratorApertureMode {
    fn from_raw(raw: &str) -> Self {
        match raw {
            "clean_aperture" => Self::CleanAperture,
            "production_aperture" => Self::ProductionAperture,
            "encoded_pixels" => Self::EncodedPixels,
            other => Self::Unknown(other.to_owned()),
        }
    }

    fn as_raw(&self) -> &str {
        match self {
            Self::CleanAperture => "clean_aperture",
            Self::ProductionAperture => "production_aperture",
            Self::EncodedPixels => "encoded_pixels",
            Self::Unknown(raw) => raw,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum AssetImageGeneratorDynamicRangePolicy {
    ForceSdr,
    MatchSource,
    Unknown(String),
}

impl AssetImageGeneratorDynamicRangePolicy {
    fn from_raw(raw: &str) -> Self {
        match raw {
            "force_sdr" => Self::ForceSdr,
            "match_source" => Self::MatchSource,
            other => Self::Unknown(other.to_owned()),
        }
    }

    fn as_raw(&self) -> &str {
        match self {
            Self::ForceSdr => "force_sdr",
            Self::MatchSource => "match_source",
            Self::Unknown(raw) => raw,
        }
    }
}

#[derive(Debug)]
pub struct AssetImageGenerator {
    ptr: *mut c_void,
}

impl Drop for AssetImageGenerator {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::av_ns_object_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

#[derive(Debug)]
pub struct AssetImage {
    ptr: *mut c_void,
}

impl Drop for AssetImage {
    fn drop(&mut self) {
        if !self.ptr.is_null() {
            unsafe { ffi::av_ns_object_release(self.ptr) };
            self.ptr = ptr::null_mut();
        }
    }
}

#[derive(Debug)]
pub struct GeneratedAssetImage {
    pub image: AssetImage,
    pub actual_time: Time,
}

impl Asset {
    pub fn image_generator(&self) -> AssetImageGenerator {
        let ptr = unsafe { ffi::av_asset_image_generator_create(self.ptr) };
        AssetImageGenerator { ptr }
    }
}

impl UrlAsset {
    pub fn image_generator(&self) -> AssetImageGenerator {
        self.asset.image_generator()
    }
}

impl AssetImageGenerator {
    fn info(&self) -> Result<AssetImageGeneratorInfoPayload, AVPlayerError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe { ffi::av_asset_image_generator_info_json(self.ptr, &mut err) };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    pub fn applies_preferred_track_transform(&self) -> Result<bool, AVPlayerError> {
        Ok(self.info()?.applies_preferred_track_transform)
    }

    pub fn set_applies_preferred_track_transform(&self, applies: bool) {
        unsafe {
            ffi::av_asset_image_generator_set_applies_preferred_track_transform(self.ptr, applies);
        }
    }

    pub fn maximum_size(&self) -> Result<Size, AVPlayerError> {
        Ok(self.info()?.maximum_size)
    }

    pub fn set_maximum_size(&self, size: Size) {
        unsafe {
            ffi::av_asset_image_generator_set_maximum_size(self.ptr, size.width, size.height);
        }
    }

    pub fn aperture_mode(&self) -> Result<Option<AssetImageGeneratorApertureMode>, AVPlayerError> {
        Ok(self
            .info()?
            .aperture_mode
            .as_deref()
            .map(AssetImageGeneratorApertureMode::from_raw))
    }

    pub fn set_aperture_mode(
        &self,
        mode: Option<AssetImageGeneratorApertureMode>,
    ) -> Result<(), AVPlayerError> {
        let mode = mode
            .map(|mode| {
                CString::new(mode.as_raw()).map_err(|error| {
                    AVPlayerError::InvalidArgument(format!(
                        "aperture mode contains NUL byte: {error}"
                    ))
                })
            })
            .transpose()?;
        unsafe {
            ffi::av_asset_image_generator_set_aperture_mode(
                self.ptr,
                mode.as_ref().map_or(ptr::null(), |mode| mode.as_ptr()),
            );
        }
        Ok(())
    }

    pub fn requested_time_tolerance_before(&self) -> Result<Time, AVPlayerError> {
        Ok(self.info()?.requested_time_tolerance_before)
    }

    pub fn set_requested_time_tolerance_before(&self, tolerance: Time) {
        let (value, timescale, kind) = tolerance.to_raw();
        unsafe {
            ffi::av_asset_image_generator_set_requested_time_tolerance_before(
                self.ptr, value, timescale, kind,
            );
        }
    }

    pub fn requested_time_tolerance_after(&self) -> Result<Time, AVPlayerError> {
        Ok(self.info()?.requested_time_tolerance_after)
    }

    pub fn set_requested_time_tolerance_after(&self, tolerance: Time) {
        let (value, timescale, kind) = tolerance.to_raw();
        unsafe {
            ffi::av_asset_image_generator_set_requested_time_tolerance_after(
                self.ptr, value, timescale, kind,
            );
        }
    }

    pub fn dynamic_range_policy(
        &self,
    ) -> Result<Option<AssetImageGeneratorDynamicRangePolicy>, AVPlayerError> {
        Ok(self
            .info()?
            .dynamic_range_policy
            .as_deref()
            .map(AssetImageGeneratorDynamicRangePolicy::from_raw))
    }

    pub fn set_dynamic_range_policy(
        &self,
        policy: &AssetImageGeneratorDynamicRangePolicy,
    ) -> Result<(), AVPlayerError> {
        let policy = CString::new(policy.as_raw()).map_err(|error| {
            AVPlayerError::InvalidArgument(format!(
                "dynamic-range policy contains NUL byte: {error}"
            ))
        })?;
        let mut err: *mut c_char = ptr::null_mut();
        let status = unsafe {
            ffi::av_asset_image_generator_set_dynamic_range_policy(
                self.ptr,
                policy.as_ptr(),
                &mut err,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { from_swift(status, err) });
        }
        Ok(())
    }

    pub fn has_custom_video_compositor(&self) -> Result<bool, AVPlayerError> {
        Ok(self.info()?.has_custom_video_compositor)
    }

    pub fn custom_video_compositor_class_name(&self) -> Result<Option<String>, AVPlayerError> {
        Ok(self.info()?.custom_video_compositor_class_name)
    }

    pub fn copy_image_at_time(
        &self,
        requested_time: Time,
    ) -> Result<GeneratedAssetImage, AVPlayerError> {
        let (value, timescale, kind) = requested_time.to_raw();
        let mut actual_value = 0_i64;
        let mut actual_timescale = 0_i32;
        let mut actual_kind = 1_i32;
        let mut err: *mut c_char = ptr::null_mut();
        let image_ptr = unsafe {
            ffi::av_asset_image_generator_copy_image_at_time(
                self.ptr,
                value,
                timescale,
                kind,
                &mut actual_value,
                &mut actual_timescale,
                &mut actual_kind,
                &mut err,
            )
        };
        if image_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        Ok(GeneratedAssetImage {
            image: AssetImage { ptr: image_ptr },
            actual_time: Time::from_raw(actual_value, actual_timescale, actual_kind),
        })
    }

    pub fn cancel_all_image_generation(&self) {
        unsafe { ffi::av_asset_image_generator_cancel_all_image_generation(self.ptr) };
    }
}

impl AssetImage {
    fn info(&self) -> Result<AssetImagePayload, AVPlayerError> {
        let mut err: *mut c_char = ptr::null_mut();
        let json_ptr = unsafe { ffi::av_asset_image_info_json(self.ptr, &mut err) };
        if json_ptr.is_null() {
            return Err(unsafe { from_swift(ffi::status::OPERATION_FAILED, err) });
        }
        parse_json_and_free(json_ptr)
    }

    pub fn width(&self) -> Result<usize, AVPlayerError> {
        Ok(self.info()?.width)
    }

    pub fn height(&self) -> Result<usize, AVPlayerError> {
        Ok(self.info()?.height)
    }

    pub fn bits_per_component(&self) -> Result<usize, AVPlayerError> {
        Ok(self.info()?.bits_per_component)
    }

    pub fn bits_per_pixel(&self) -> Result<usize, AVPlayerError> {
        Ok(self.info()?.bits_per_pixel)
    }

    pub fn bytes_per_row(&self) -> Result<usize, AVPlayerError> {
        Ok(self.info()?.bytes_per_row)
    }

    pub fn alpha_info(&self) -> Result<u32, AVPlayerError> {
        Ok(self.info()?.alpha_info)
    }

    pub fn bitmap_info(&self) -> Result<u32, AVPlayerError> {
        Ok(self.info()?.bitmap_info)
    }

    pub fn rendering_intent(&self) -> Result<u32, AVPlayerError> {
        Ok(self.info()?.rendering_intent)
    }
}
